File: C++/92. Reverse Linked List II.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *find(ListNode *head, int pos)
{
    while (--pos > 0)
        head = head->next;

    return head;
}

ListNode *reverse(ListNode *head, ListNode *end)
{
    ListNode *cur = head, *prev = end->next, *next;

    while (cur and cur != end)
    {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }

    if (cur)
        cur->next = prev;

    return cur;
}

ListNode *reverseBetween(ListNode *head, int left, int right)
{
    ListNode *lNode = find(head, left - 1);          // Finding Prev of Left
    ListNode *rNode = find(lNode, right - left + 1); // Finding right

    if (left == 1)
        return reverse(head, rNode);

    lNode->next = reverse(lNode->next, rNode->next);

    return head;
}

File: Java/92. Reverse Linked List II.java
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode start = findNodeByPosition(head, left - 1);
        ListNode end = findNodeByPosition(start, right - left + 1);

        if (left == 1)
            return reverseNode(head, end);

        start.next = reverseNode(start.next, end.next);

        return head;
    }

    private ListNode reverseNode(ListNode head, ListNode end) {
        ListNode prev = end.next, next;

        while (head != end) {
            next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }

        if (head != null)
            head.next = prev;

        return head;
    }

    private ListNode findNodeByPosition(ListNode head, int pos) {
        while (--pos > 0)
            head = head.next;

        return head;
    }

    /* ignore Boiler Code */
    public class ListNode {
        public int val;
        public ListNode next;

        public ListNode() {
        }

        public ListNode(int val) {
            this.val = val;
        }

        public ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}